### ISSUE TITLE
Some failed pods in add periodic update may become running in the process

### ISSUE THREAD

**Issue by Mmduh-483** [Mellanox/ib-kubernetes#65]
If multus didn't find the network CRD that a pod requested, then multus just ignore that CRD without failing. for the following pod:
```
...
annotations:
  k8s.v1.cni.cncf.io/networks: [{"name": "my-network"}]
...
```
if the network CRD `my-network` is not created, then multus will ignore it, so the pod will be running, while it is stuck in the ib-kubernetes add periodic update, even if the network created later, and the pod get configured, it will be late since the pod is running.

**Comment by Mmduh-483** [comment]
@moshe010 @adrianchiris

**Comment by DmytroLinkin** [comment]
Pod's network annotation:
```
...
  annotations:
    k8s.v1.cni.cncf.io/networks: some-network
...
```
After creating pod it hangs at `ContainerCreating` step.
```
$ kubectl describe pod failed-pod
...
  Warning  FailedCreatePodSandBox  0s (x4 over 6s)    kubelet, hostname (combined from similar events): 
Failed to create pod sandbox: rpc error: code = Unknown desc = failed to set up sandbox container 
"a1cf080aa5630453c5044803c97578bfb5ab2e9b6585a9c067ed39cddce16f45" network for pod "failed-pod": 
networkPlugin cni failed to set up pod "failed-pod_default" network: Multus: [default/failed-pod]: error loading 
k8s delegates k8s args: TryLoadPodDelegates: error in getting k8s network from pod: GetNetworkDelegates: 
failed getting the delegate: getKubernetesDelegate: cannot find get a network-attachment-definition
 (some-network) in namespace (default): network-attachment-definitions.k8s.cni.cncf.io "some-network" not found
```

**Comment by adrianchiris** [comment]
I dont quite understand the issue @Mmduh-483,
The problem was with multus according to the original description, am i right ?

@DmytroLinkin please try the alternate form for setting the network annotation and see if it behaves the same.
i.e `k8s.v1.cni.cncf.io/networks: [{"name": "some-network"}]`

**Comment by DmytroLinkin** [comment]
> @DmytroLinkin please try the alternate form for setting the network annotation and see if it behaves the same.
> i.e `k8s.v1.cni.cncf.io/networks: [{"name": "some-network"}]`

Creating pod with such annotation lead to
```
error: error validating "failed-pod": error validating data: ValidationError(Pod.metadata.annotations.k8s.v1.cni.cncf.io/networks): invalid type for io.k8s.apimachinery.pkg.apis.meta.v1.Objec
tMeta.annotations: got "array", expected "string"; if you choose to ignore these errors, turn validation off with --validate=false
```

**Comment by DmytroLinkin** [comment]
@adrianchiris, my bad, didn't notice to put quotes around braces, but error same as in the first comment

**Comment by Mmduh-483** [comment]
@adrianchiris @DmytroLinkin using multus without specifying the `kubeconfig` field for multus config, then multus will not try to get the networkattachments that the pod specified in the annotations, but it will stuck in the periodic update of ib-kubernetes.

1- Here multus ignore the kubeclient if it is null https://github.com/intel/multus-cni/blob/master/multus/multus.go#L510 
2- Here multus assume it should return a kubeclient as null if there is no `kubeconfig` https://github.com/intel/multus-cni/blob/c85b79f5ff5bcacaa45e2135d29e9afb6b84ed9b/k8sclient/k8sclient.go#L399

**Comment by DmytroLinkin** [comment]
> @adrianchiris @DmytroLinkin using multus without specifying the `kubeconfig` field for multus config, then multus will not try to get the networkattachments that the pod specified in the annotations, but it will stuck in the periodic update of ib-kubernetes.

Than it's a misconfiguration?

**Comment by adrianchiris** [comment]
Its a miss configuration for kubernetes infiniband case.

we should IMO:
1. document this so its clear this is a requirement to configure multus this way
2. introduce a retry or aging mechanism which eventually return an error for a pod in case an API request failed

there are several endpoints we are accessing:
1. kube-apiserver
2. ufm

I think we should do for both as it would be the same mechanism, an API call may fail due to traffic issues or issues with the server (temporary/permanent)

**Comment by adrianchiris** [comment]
this issue was addressed in #82 and #84
